python-api/tests/integrationtests/test_controller.py

```diff
@@ -3,725 +3,343 @@
+        set_composite_mode is called multiple times with the same format
+        """
+        self.setup_server()
+        self.setup_controller()
+
+        self.log.info("setting callback")
+        test_cb = Mock(side_effect=self.quit_mainloop)
+        self.controller.on_new_mode_online(test_cb)
+
+        self.setup_sources()
+
+        self.log.info("waiting for initial callback with"
+                      "default-mode COMPOSITE_DUAL_EQUAL")
+        self.run_mainloop(timeout=5)
+        test_cb.assert_called_once_with(Controller.COMPOSITE_DUAL_EQUAL)
+        test_cb.reset_mock()
+
+        self.log.info("setting the same composite mode"
+                      "(COMPOSITE_DUAL_EQUAL) again")
+        assert not self.controller.set_composite_mode(
+            Controller.COMPOSITE_DUAL_EQUAL)
+
+        self.log.info("setting a new composite-mode COMPOSITE_PIP")
+        assert self.controller.set_composite_mode(Controller.COMPOSITE_PIP)
+        self.run_mainloop(timeout=5)
+
+        # just waiting for the timeout to verify no incoming call
+        # would slow down the tests remarebly, so we send another
+        # mode-change, knowing that when it arrives without an
+        # intermediate call, there was no duplicate callback
+
+        self.log.info("waiting for callback with new mode COMPOSITE_PIP")
+        test_cb.assert_called_once_with(Controller.COMPOSITE_PIP)
+
+    def test_preview_port_added_cb(self):
+        """Create a Controller object, call on_preview_port_added method and
+        check that the callback fires correctly when new video-sources
+        are added
+        """
+        self.setup_server()
+        self.setup_controller()
+
+        test_cb = Mock(side_effect=lambda mode, serve, type:
+                       self.quit_mainloop_after(call_count=2))
+        self.controller.on_preview_port_added(test_cb)
+
+        self.setup_sources()
+        self.run_mainloop(timeout=5)
+
+        self.log.info("test_cb called with: %s", test_cb.call_args_list)
+        test_cb.assert_any_call(3003, 1, 7)
+        test_cb.assert_any_call(3004, 1, 8)
+        assert test_cb.call_count == 2
+
+
+class TestNewRecord(IntegrationTestbase):
+    """ Test new_record method
+    """
+
+    def wait_for_sources(self, count):
+        """ Blocks until the Server has reported, that the right number of
+        preview-ports are is started
+        """
+        self.log.info("waiting for Server to start preview-port-outputs")
+        self.serv.wait_for_output('tcpserversink name=sink', count=count)
+
+    def setup_sources(self):
+        """ Starts some Test-Video streams and waits until they are ready
+        """
+        self.log.info("starting 2 test-video sources")
+        self.new_test_video()
+        self.new_test_video()
+
+        self.log.info("waiting for the test-video sources to come up")
+        self.wait_for_sources(count=2)
+
+    def test_filename_changes(self):
```

Comment:
No, we have no such signal. I'm planning to at two other signals after this PR, I'm adding this to my TODO List